### PR TITLE
 Added typing annotations to transforms/functional_pil

### PR DESCRIPTION
**Pull request by oke-aditya**
Following up on #2025, this PR adds missing typing annotations in functional_pil.py

Any feedback is welcome!

cc @frgfm @pmeier

### CODE REVIEW

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:131]: Not sure if `List[int]` would suffice.

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:132]: A bit unsure about this too.

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:295]: I'm not sure what matrix is. I think it's `Any`

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:329]: How do we type interpolation :question:

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:374]: Please verify. I think it's int. (after digging into PIL code)

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:367]: Please verify. I think it's int. (after digging into PIL code)

**Review by oke-aditya — COMMENTED** [review]
Little help would be nice! I'm not sure about few

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [inline torchvision/transforms/functional_pil.py:227]:
```suggestion
    size: Union[Sequence[int], int],
```

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [inline torchvision/transforms/functional_pil.py:228]:
```suggestion
    interpolation: int = Image.BILINEAR,
```

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [inline torchvision/transforms/functional_pil.py:270]:
```suggestion
    fill: Union[numbers.Number, List[numbers.Number], Tuple[numbers.Number, ...]],
```

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [reply torchvision/transforms/functional_pil.py:329]: As `int`. See https://github.com/python-pillow/Pillow/blob/f518149a1bbe5ebea9d0ca41612e3e6748349ace/src/PIL/Image.py#L184-L190

```suggestion
    interpolation: int = Image.BICUBIC,
```

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [inline torchvision/transforms/functional_pil.py:315]: Please apply the same as I suggested above here and anywhere else.

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [reply torchvision/transforms/functional_pil.py:367]: Yep, [seems to be good](https://pillow.readthedocs.io/en/stable/reference/ImageOps.html#PIL.ImageOps.posterize):

> bits – The number of bits to keep for each channel (1-8).

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [reply torchvision/transforms/functional_pil.py:374]: [Seems to be good](https://pillow.readthedocs.io/en/stable/reference/ImageOps.html?highlight=posterize#PIL.ImageOps.solarize)

> threshold – All pixels above this greyscale level are inverted.

Although I'm not sure if `float` would also be possible for floating point images (mode `F`).

**Review by pmeier — COMMENTED** [review]
Hey @oke-aditya and sorry for the delay. I have some comments inline.

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [inline torchvision/transforms/functional_pil.py:110]:
```suggestion
    gain: float = 1.0,
```

**Review by oke-aditya — COMMENTED** [review]
I think I have fixed the bugs.
Should we unblock this file from mypy.ini ?

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:315]: Here

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:330]: Here

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [inline torchvision/transforms/functional_pil.py:297]: Here

**Review by oke-aditya — COMMENTED** [review]
Just marking the lines where error occured

**Comment by oke-aditya** [comment]
Hmmm tests seem to be failing on `.pad()` function. 
Not very sure why it occured, but I think that has easy fix by passing args as `padding=pad`.

Also another point I noticed that. 

The functional tensor has similar [padding code](https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional_tensor.py#L410), and I think that it accepts `float`, `int` as well as List and Tuples of them.
But the typing seems to be only `List[int]`.

I think that the typing for Functional PIL and Functional Tensor should be same. Since that's how we [test them](https://github.com/pytorch/vision/blob/master/test/test_functional_tensor.py#L849) as well.

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [inline torchvision/transforms/functional_pil.py:132]: 1. `mypy` treats `int` as virtual subtype of `float`. So if we substitute `numbers.Number` with `float`, we should be good even if we set the default argument to `0`.
2. Tests fail here, because we missed the ` = 0` default argument.

**Review by pmeier — COMMENTED** [review]
> We have stumbled upon [mypy/#3186](https://github.com/python/mypy/issues/3186) issue because of which mypy still complains.
> It thinks `int` is not a number and hence we cannot assign `0` as default arg.

Good catch. I've added a fix inline.

> Hmmm tests seem to be failing on `.pad()` function.
> Not very sure why it occured, but I think that has easy fix by passing args as `padding=pad`.

Fix inline.

> The functional tensor has similar [padding code](https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional_tensor.py#L410), and I think that it accepts `int` as well as List and Tuples of them.
> (Basically same as `Union[int, List[int], Tuple[int, ...]],`)
> 
> But the typing seems to be only `List[int]`.
> 
> I think that the typing for Functional PIL and Functional Tensor should be same. Since that's how we [test them](https://github.com/pytorch/vision/blob/master/test/test_functional_tensor.py#L849) as well.

Yes the signatures and in turn the annotations should be the same. Would you send a follow-up PR after this one is merged?

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [reply torchvision/transforms/functional_pil.py:132]: Yeah I just spotted a minute ago too. :smile: (Coincidence)

I will cover these small fixes up and in a follow up PR ensure that the annotations between `functional_pil` and `functional_tesnor ` match properly. Not just on the pad case.

**Comment by oke-aditya** [comment]
@datumbox @pmeier 
I think so CI is green enough to merge this. Probably all the failures are unrelated.

Hopefully, Good to go!

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [inline torchvision/transforms/functional_pil.py:132]: This is defined as int on the tensor equivalent method, though it supports both. Do we want them aligned?

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [inline torchvision/transforms/functional_pil.py:227]: The tensor API uses List instead of Sequence. Do we want them aligned?

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [inline torchvision/transforms/functional_pil.py:270]: This is also affected if we turn the above to int.

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [inline torchvision/transforms/functional_pil.py:297]: FYI strangely the tensor API defines this as float as well.

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [inline torchvision/transforms/functional_pil.py:296]: Perhaps we should us Image.NEAREST instead of the numeric here?

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [inline torchvision/transforms/functional_pil.py:312]: same here

**Review by datumbox — COMMENTED** [review]
@oke-aditya I checked this again. Left a couple of comments to get your input. It's mainly around whether we want to align the API between PIL and Tensor or if we want to keep the two separate. No strong opinions here. @pmeier Let me also know what you think please.

I know this PR turned out to be lengthy so if you had enough we can address them later. Cheers! :)

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [reply torchvision/transforms/functional_pil.py:132]: I think we want float in tensor equivalent method to align this.

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [reply torchvision/transforms/functional_pil.py:227]: Yes, I think we need to have Sequence in tensor, to align. But probably JIT does not support it.

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [reply torchvision/transforms/functional_pil.py:296]: Hmm, something stranger is that in the tensor API we pass a string.
I'm not sure how would we align it. (again probably JIT?)

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [reply torchvision/transforms/functional_pil.py:297]: Actually the tensor API defines it to be only `List[float]` and not all the three, again something which isn't aligned or probably has an issue.

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [reply torchvision/transforms/functional_pil.py:312]: We could change it as you suggested, but again this is inconsistent with functional_tensor.

**Review by oke-aditya — COMMENTED** [review]
Hey @datumbox ! Here are my thoughts.

There are probably multiple places where we have misaligned with functional_tensor.

My idea was to follow-up this with a new PR. (#4282 ).


Let me know what you think!

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [reply torchvision/transforms/functional_pil.py:132]: no strong opinion here. I leave it up to you.

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [reply torchvision/transforms/functional_pil.py:297]: It's the union that is not supported. I just highlight that here the type (float vs int) is actually aligned. Which is weird cause we don't seem to keep things aligned. I probably agree that it's the tensor that needs changing though.

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [reply torchvision/transforms/functional_pil.py:312]: So the reason why I mention this is because on other methods in this file you use the "enum" while in this case you use the integer equivalent (see `perspective` for example). I think the enum will be clearer.

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [reply torchvision/transforms/functional_pil.py:296]: I mean something different. See the next comment.

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [reply torchvision/transforms/functional_pil.py:312]: Aah cool, I get it. I will rewrite with Enum :smile:

Comment (datumbox) [reply torchvision/transforms/functional_pil.py:312]: yeah sorry I should have been clearer. OK ping me when this is done and I'll approve. Thanks for the effort and sorry for the back and forth. :)

**Comment by oke-aditya** [comment]
I think this should be do it! @datumbox can you just cross check if I have kept the modes `BILINEAR` `CUBIC` and `NEAREST` correct as per old definitions?

File: torchvision/transforms/functional_pil.py
Comment (oke-aditya) [reply torchvision/transforms/functional_pil.py:297]: I just verified Sequence is not supported by JIT. And yes Union too is not supported.

https://pytorch.org/docs/stable/jit_language_reference.html#supported-type

**Review by datumbox — APPROVED** [review]
LGTM, thank you @oke-aditya for your contribution and for your patience. :) 

I checked the modes and they look fine to me.

@pmeier Do you mind also having a look that we are good to go? Once I get your stamp we can merge.

File: torchvision/transforms/functional_pil.py
Comment (pmeier) [reply torchvision/transforms/functional_pil.py:227]: Yes, `torch.jit` does not support `Sequence`. Since we also have a `Union` here, we can't align them anyway. Thus, I think it is ok to go with `Sequence`.

**Review by pmeier — APPROVED** [review]
> can you just cross check if I have kept the modes `BILINEAR` `CUBIC` and `NEAREST` correct as per old definitions?

Seems to be good, even replacing an old `0` with `NEAREST` :+1:

LGTM, thanks a lot @oke-aditya!

**Pull request merged**
